src/Acuminator/Acuminator.Utils/RoslynExtensions/NestedInvocationWalker.cs

```diff
@@ -0,0 +1,156 @@
+	/// string descr = SomeHelper.GetDescription(); // this code is being analyzed
+	/// ...
+	/// // In some other file or even in a different assembly
+	/// public static class SomeHelper
+	/// {
+	///		public static void GetDescription()
+	///		{
+	///			var graph = new PXGraph(); // this code will be analyzed too
+	///		}
+	/// }
+	///	</code>
+	/// </example>
+	// ReSharper disable once InheritdocConsiderUsage
+	public class NestedInvocationWalker : CSharpSyntaxWalker
+	{
+		private const int MaxDepth = 100; // to avoid circular dependencies
+
+		private readonly SemanticModel _semanticModel;
+		private CancellationToken _cancellationToken;
+
+		private SyntaxNode _currentNode;
+		private SyntaxNode _originalNode;
+
+		private readonly CodeAnalysisSettings _settings;
+
+		/// <summary>
+		/// Syntax node in the original tree that is being analyzed.
+		/// Typically it is the node on which a diagnostic should be reported.
+		/// </summary>
+		protected SyntaxNode OriginalNode => _originalNode ?? _currentNode;
+
+		private readonly Stack<SyntaxNode> _nodesStack = new Stack<SyntaxNode>();
+
+		public NestedInvocationWalker(SemanticModel semanticModel, CancellationToken cancellationToken)
+		{
+			semanticModel.ThrowOnNull(nameof (semanticModel));
+
+			_semanticModel = semanticModel;
+			_cancellationToken = cancellationToken;
+
+			try
+			{
+				if (ServiceLocator.IsLocationProviderSet)
+					_settings = ServiceLocator.Current.GetInstance<CodeAnalysisSettings>();
+			}
+			catch
+			{
+				// TODO: log the exception
+			}
+
+			if (_settings == null)
+				_settings = CodeAnalysisSettings.Default;
+		}
+
+		protected void ThrowIfCancellationRequested()
+		{
+			_cancellationToken.ThrowIfCancellationRequested();
+		}
+
+		/// <summary>
+		/// Returns a symbol for an invocation expression, or, 
+		/// if the exact symbol cannot be found, returns the first candidate.
+		/// </summary>
+		protected IMethodSymbol GetMethodSymbol(InvocationExpressionSyntax node)
+		{
+			var symbolInfo = _semanticModel.GetSymbolInfo(node, _cancellationToken);
+
+			if (symbolInfo.Symbol is IMethodSymbol methodSymbol)
+			{
+				return methodSymbol;
+			}
+
+			if (!symbolInfo.CandidateSymbols.IsEmpty)
+			{
+				return symbolInfo.CandidateSymbols.OfType<IMethodSymbol>().FirstOrDefault();
+			}
+
+			return null;
+		}
+
+		private void Push(SyntaxNode node)
+		{
+			if (_nodesStack.Count == 0)
+				_originalNode = node;
+
+			_nodesStack.Push(node);
+		}
+
+		private void Pop()
+		{
+			_nodesStack.Pop();
+
+			if (_nodesStack.Count == 0)
+				_originalNode = null;
+		}
+
+		private bool RecursiveAnalysisEnabled()
+		{
+			return _settings.RecursiveAnalysisEnabled && _nodesStack.Count <= MaxDepth;
+		}
+
+		#region Visit
+
+		public override void DefaultVisit(SyntaxNode node)
+		{
+			_currentNode = node;
+			base.DefaultVisit(node);
+			_currentNode = null;
+		}
+
+		public override void VisitInvocationExpression(InvocationExpressionSyntax node)
```

Comment:
@dropsonic just the invocation expressions isn't enough for analysis - what about property access. It doesn't contain InvocationExpressionSyntax  node:
`CodeAnalysisSettings.Default` for example

Reply:
@SENya1990 it doesn't contain many features like lambdas, property getters, etc.; this is an initial draft